frontend/src/metabase/dashboard/actions/tabs.ts

```diff
@@ -309,6 +318,24 @@ export const tabsReducer = createReducer<DashboardState>(
         }
         const dashCard = state.dashcards[dashCardId];
 
+        if (!dashCard.dashboard_tab_id) {
+          throw Error(
+            `MOVE_DASHCARD_TO_TAB was dispatched but dashCard.dashboard_tab_id (${dashCard.dashboard_tab_id}) is null`,
+          );
+        }
```

Comment:
Is there a case where this could happen?

Reply:
It shouldn't but typescript doesn't know it, we don't allow `dashboard_tab_id!` so I added this assertion, do we have another pattern for this?

Reply:
Fair enough. I think `dashboard_tab_id` could be `null` when there's no tab in the dashboard as well. Indeed, it won't be the case obviously if `moveDashCardToTab` is called there're gotta be tabs. But I dislike the fact that we have to satisfy TypeScript and add guard blocks unnecessarily. It clutters the code and could give the impression that this could actually happen.

I'm not aware of other patterns but maybe we could cast `dashCard` to a new tab that has `dashboard_tab_id` as a number here?

@oisincoveney @uladzimirdev @kamilmielnik @deniskaber are you guys aware of patterns we do in our code to deal with this situation?

Reply:
We use [`checkNotNull`](https://github.com/metabase/metabase/blob/8c0313dba0226b75f8f2d6530c0d2d3aee6f1b9f/frontend/src/metabase/core/utils/types/types.ts#L11).

Reply:
👆 That works. We could extract `dashboard_tab_id` to another variable, but it's a shame we couldn't just call it from `dashcard` anymore.

```ts
const dashboardTabId = checkNotNull(dashcard.dashboard_tab_id)
```

Reply:
Ah yep `checkNotNull` is the way to go. My other question then is - are we happy introducing a generic `TypeError` without a message? I like the fact that the error message is specific here.

Reply:
<img width="574" alt="image" src="https://github.com/metabase/metabase/assets/25306947/e4fdd125-eab7-485a-a98a-b5bc18206c92">

Ofc we have the stack trace so it's not a problem in terms of finding the problem

Reply:
What do you think about an "assert" function like this?
```ts
export const assertNotNull = <T>(
  value: T | null | undefined,
  message?: string,
): asserts value is T => {
  if (value == null) {
    throw new TypeError(message);
  }
};

```

It wouldn't force us to make new variables and we can just put the assertions at the start of the function, in this case:
```diff
builder.addCase(
      _moveDashCardToTab,
      (state, { payload: { dashCardId, destinationTabId } }) => {
+       assertNotNull(dashCard.dashboard_tab_id);
        const { dashId } = getPrevDashAndTabs({ state });
+		assertNotNull(dashId);
-        if (dashId === null) {
-          throw Error(
-            `MOVE_DASHCARD_TO_TAB was dispatched but dashId (${dashId}) is null`,
-          );
-        }
        const dashCard = state.dashcards[dashCardId];

-        if (!dashCard.dashboard_tab_id) {
-          throw Error(
-            `MOVE_DASHCARD_TO_TAB was dispatched but dashCard.dashboard_tab_id -(${dashCard.dashboard_tab_id}) is null`,
-          );
-        }

        const dashboard = state.dashboards[dashId];

        if (!dashboard.dashCardTabMovements) {
```

Reply:
I also read "assert not null" more as "I expect this to not be null, if it isn't, something is broken", while "check not null" doesn't give me the same idea

Reply:
> What do you think about an "assert" function like this?

This might be good candidate for an RFC. For now `checkNotNull` will do.